src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,976 @@
+                            SyntaxFactory.LiteralExpression(SyntaxKind.NumericLiteralExpression, SyntaxFactory.Literal(0)), // count = 0
+                            variableName: "count",
+                            out documentUpdateInfo);
+                }
+
+                documentUpdateInfo = default;
+                return false;
+            }
+
+            private bool TryConvertIfInToListInvocation(
+                InvocationExpressionSyntax invocationExpression,
+                QueryExpressionProcessingInfo queryExpressionProcessingInfo,
+                out DocumentUpdateInfo documentUpdateInfo)
+            {
+                // before var list = (from a in b select a).ToList();
+                // after
+                // var list = new List<T>();
+                // foreach (var a in b)
+                // {
+                //     list.Add(a)
+                // }
+                if (_semanticModel.GetSymbolInfo(invocationExpression, _cancellationToken).Symbol is IMethodSymbol methodSymbol &&
+                    methodSymbol.RefKind == RefKind.None &&
+                    IsList(methodSymbol.ReturnType) &&
+                    methodSymbol.Parameters.Length == 0)
+                {
+                    return TryConvertIfInInvocation(
+                              invocationExpression,
+                              queryExpressionProcessingInfo,
+                              IsList,
+                              (listIdentifier, expression) => SyntaxFactory.ExpressionStatement(SyntaxFactory.InvocationExpression(
+                                    SyntaxFactory.MemberAccessExpression(
+                                        SyntaxKind.SimpleMemberAccessExpression,
+                                        listIdentifier,
+                                        SyntaxFactory.IdentifierName(nameof(IList.Add))),
+                                    SyntaxFactory.ArgumentList(SyntaxFactory.SingletonSeparatedList(SyntaxFactory.Argument(expression))))),
+                              SyntaxFactory.ObjectCreationExpression(
+                                  methodSymbol.GenerateReturnTypeSyntax().WithAdditionalAnnotations(Simplifier.Annotation),
+                                  SyntaxFactory.ArgumentList(),
+                                  initializer: null),
+                               variableName: "list",
+                              out documentUpdateInfo);
+                }
+
+                documentUpdateInfo = default;
+                return false;
+            }
+
+            private bool IsInt(ITypeSymbol typeSymbol)
+                => typeSymbol.SpecialType == SpecialType.System_Int32;
+
+            private bool IsList(ITypeSymbol typeSymbol)
+                => Equals(typeSymbol.OriginalDefinition, _semanticModel.Compilation.GetTypeByMetadataName(typeof(List<>).FullName));
```

Comment:
>Equals(typeSymbol.OriginalDefinition, _semanticModel.Compilation.GetTypeByMetadataName(typeof(List<>).FullName)); [](start = 19, length = 113)

Do we understand why ```SymbolEquivalenceComparer.Instance``` didn't work for us? Was it not supposed to? Let's put this on the list to follow up, unless you have an answer.

Reply:
Thank you! Added to [https://github.com/dotnet/roslyn/issues/25639](https://github.com/dotnet/roslyn/issues/25639)

---
In reply to: [178432101](https://github.com/dotnet/roslyn/pull/25362#discussion_r178432101) [](ancestors = 178432101)